### PR TITLE
TextToSpeech coverage

### LINKED ISSUE #4810: Add unit test for KiwixTextToSpeech
**Type of Test**

* [x] Unit Test
* [ ] UI Test

**Particular area and files to be tested**
kiwix-android/core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

**Current Status of the tests**

Image

No direct unit tests currently exist for KiwixTextToSpeech

**Particular Behaviour to be tested**

- Verify unsupported or missing TTS language shows the proper toast and does not start reading.
- Verify `initializeTTS()` marks TTS as initialized and calls the onInitSucceed callback.
- Verify failed initialization shows the proper toast and does not call the success callback.
- Verify `initWebView()` adds the tts JavaScript interface to the WebView.
- Verify `readAloud()` clears a paused task and triggers `onSpeakingEnded()`.
- Verify blank or separator only content does not create a TTS task or start speaking.
- Verify language "mul" shows the “read aloud not enabled” toast and does not start reading.

### PR DESCRIPTION
**Pull request by Piyush-Kumar-Mishra**
FIX #4810

### CODE REVIEW

**Comment by kelson42** [comment]
@Piyush-Kumar-Mishra I guess this is ready for review?

**Comment by Piyush-Kumar-Mishra** [comment]
@kelson42 Yes, ready for review

File: core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+  fun `readAloud does not speak when language is multilingual`() {
+    injectMockTts()
+    every { tts.isSpeaking } returns false
+    every { zimReaderContainer.language } returns "mul"
+    kiwixTts.readAloud(webView)
+    verify(exactly = 0) { webView.loadUrl(match { it.contains("body") }) }
+  }
+
+  @Test
+  fun `readAloud does not speak when language is not supported`() {
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt:140]: These tests are good. But there are some more scenarios that should be tested.

* Read aloud should run when everything is valid.
* Read aloud does not run when audio focus is denied.
* Read aloud does not run when the locale is null.
* Read aloud shows the TTSLanguageDownloadDialog.


```kotlin
 @Test
  fun `readAloud loads processed content when everything is valid`() {
    injectMockTts()
    grantAudioFocus()
    kiwixTts.readAloud(setupReadAloudForSuccess())
    verify {
      webView.loadUrl(match { it.contains("body = document.getElementsByTagName") })
    }
  }

  @Test
  fun `readAloud does not proceed when audio focus is denied`() {
    injectMockTts()
    every {
      audioManager.requestAudioFocus(any<AudioFocusRequest>())
    } returns AudioManager.AUDIOFOCUS_REQUEST_FAILED

    kiwixTts.readAloud(setupReadAloudForSuccess())

    verify(exactly = 0) { webView.loadUrl(any()) }
  }

  @Test
  fun `readAloud does not proceed when locale is null`() {
    injectMockTts()

    every { tts.isSpeaking } returns false
    every { zimReaderContainer.language } returns "eng"

    mockkObject(LanguageUtils.Companion)
    every { LanguageUtils.iSO3ToLocale(any()) } returns null

    kiwixTts.readAloud(webView)

    verify(exactly = 0) { webView.loadUrl(any()) }
  }

  @Test
  fun `readAloud shows download dialog when feature not installed`() {
    val activity = mockk<CoreMainActivity>(relaxed = true)
    every { activity.getSystemService(Context.AUDIO_SERVICE) } returns audioManager
    kiwixTts = KiwixTextToSpeech(
      activity,
      initListener,
      speakingListener,
      focusListener,
      zimReaderContainer
    )
    injectMockTts()
    every { tts.isSpeaking } returns false
    every { zimReaderContainer.language } returns "eng"
    every { tts.isLanguageAvailable(any()) } returns TextToSpeech.LANG_AVAILABLE

    val voice: Voice = mockk(relaxed = true)
    every { voice.features } returns setOf(TextToSpeech.Engine.KEY_FEATURE_NOT_INSTALLED)
    every { tts.voice } returns voice

    kiwixTts.readAloud(webView)

    verify { activity.externalLinkOpener.showTTSLanguageDownloadDialog() }
  }
```

File: core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    val task = kiwixTts.TTSTask(pieces)
+    task.start()
+    task.pause()
+    assertThat(task.paused).isTrue()
+    verify { tts.setOnUtteranceProgressListener(null) }
+    verify { tts.stop() }
+  }
+
+  @Test
+  fun `task stop clears current task and notifies listener`() {
```
Comment (MohitMaliFtechiz) [inline core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt:266]: We should test the other edge cases as well.

```kotlin
@Test
  fun `task continues speaking next piece on done`() {
    injectMockTts()

    val pieces = listOf("Hello", "World")
    val task = kiwixTts.TTSTask(pieces)

    task.start()

    val listenerSlot = slot<UtteranceProgressListener>()
    verify { tts.setOnUtteranceProgressListener(capture(listenerSlot)) }

    listenerSlot.captured.onDone("id")

    verify { tts.speak(eq("World"), any(), any(), any()) }
  }

  @Test
  fun `task handles error and shows toast`() {
    injectMockTts()

    val pieces = listOf("Hello")
    val task = kiwixTts.TTSTask(pieces)

    task.start()

    val listenerSlot = slot<UtteranceProgressListener>()
    verify { tts.setOnUtteranceProgressListener(capture(listenerSlot)) }

    listenerSlot.captured.onError("error")
    verify { tts.setOnUtteranceProgressListener(any()) }
  }

@Test
  fun `task stops after last piece`() {
    injectMockTts()
    val pieces = listOf("OnlyOne")
    val task = kiwixTts.TTSTask(pieces)
    kiwixTts.currentTTSTask = task
    task.start()
    val listenerSlot = slot<UtteranceProgressListener>()
    verify { tts.setOnUtteranceProgressListener(capture(listenerSlot)) }
    listenerSlot.captured.onDone("id")
    assertThat(kiwixTts.currentTTSTask).isNull()
    verify { speakingListener.onSpeakingEnded() }
  }

  @Test
  fun `pause decrements current piece and resumes correctly`() {
    injectMockTts()
    val pieces = listOf("Hello", "World")
    val task = kiwixTts.TTSTask(pieces)
    task.start()
    task.pause()
    task.start()
    verify(atLeast = 2) {
      tts.speak(eq("Hello"), any(), any(), any())
    }
  }

 @Test
  fun `javascript interface splits content and starts speaking`() {
    injectMockTts()

    val jsInterface = kiwixTts.javaClass
      .declaredClasses
      .first { it.simpleName == "TTSJavaScriptInterface" }
      .getDeclaredConstructor(kiwixTts.javaClass)
      .apply { isAccessible = true }
      .newInstance(kiwixTts)

    val method = jsInterface.javaClass.getDeclaredMethod("speakAloud", String::class.java)
    method.invoke(jsInterface, "Hello. World\nTest")

    verify { speakingListener.onSpeakingStarted() }
    assertThat(kiwixTts.currentTTSTask).isNotNull()
  }

  @Test
  fun `javascript interface does nothing for empty content`() {
    injectMockTts()

    val jsInterface = kiwixTts.javaClass
      .declaredClasses
      .first { it.simpleName == "TTSJavaScriptInterface" }
      .getDeclaredConstructor(kiwixTts.javaClass)
      .apply { isAccessible = true }
      .newInstance(kiwixTts)

    val method = jsInterface.javaClass.getDeclaredMethod("speakAloud", String::class.java)
    method.invoke(jsInterface, "   \n   ")

    verify(exactly = 0) { speakingListener.onSpeakingStarted() }
    assertThat(kiwixTts.currentTTSTask).isNull()
  }

```

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]
@Piyush-Kumar-Mishra, There are some changes @Copliot suggested. Apart from that, please see the review changes.

**Comment by Piyush-Kumar-Mishra** [comment]
Thank you for the review
I’ll check everything again and update the commit.

File: core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt (hunk 2 above)
Comment (Piyush-Kumar-Mishra) [reply core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt:266]: In this test -> "task handles error and shows toast"
the toast isn’t actually being tested.

Comment (MohitMaliFtechiz) [reply core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt:266]: Please verify that :)

Comment (Piyush-Kumar-Mishra) [reply core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt:266]:
```
  @Test
  fun `task handles error and shows toast`() {
    injectMockTts()

    val pieces = listOf("Hello")
    val task = kiwixTts.TTSTask(pieces)

    task.start()

    val listenerSlot = io.mockk.slot<android.speech.tts.UtteranceProgressListener>()
    verify { tts.setOnUtteranceProgressListener(capture(listenerSlot)) }
    listenerSlot.captured.onError("error")

    // Validate that the toast is actually shown
    assertThat(ShadowToast.getTextOfLatestToast())
      .isEqualTo(context.getString(org.kiwix.kiwixmobile.core.R.string.texttospeech_error))
  }

```

Comment (MohitMaliFtechiz) [reply core/src/test/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeechTest.kt:266]: Please do not use the classes, method, or string with the whole package name. Always import that and simply use that.

* `io.mockk.slot` should be `slot`.
* `android.speech.tts.UtteranceProgressListener` should be `UtteranceProgressListener`.


Same for others.

**Review by MohitMaliFtechiz — DISMISSED** [review]
There were some import issues(and commit cleanup). I have fixed both issues. @kelson42 Now, LGTM. Once CI is green, we can merge.

**Comment by kelson42** [comment]
CI is not green at all :(

**Comment by MohitMaliFtechiz** [comment]
@kelson42 We have improved the flaky test. Now the CI is green.
